File: api/src/main/java/com/readrops/api/services/freshrss/FreshRSSService.java
package com.readrops.api.services.freshrss;

import com.readrops.api.services.freshrss.json.FreshRSSUserInfo;
import com.readrops.db.entities.Feed;
import com.readrops.db.entities.Folder;
import com.readrops.db.entities.Item;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.http.Body;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Query;

public interface FreshRSSService {

    String END_POINT = "/api/greader.php/";

    @POST("accounts/ClientLogin")
    Single<ResponseBody> login(@Body RequestBody body);

    @GET("reader/api/0/token")
    Single<ResponseBody> getWriteToken();

    @GET("reader/api/0/user-info")
    Single<FreshRSSUserInfo> getUserInfo();

    @GET("reader/api/0/subscription/list?output=json")
    Single<List<Feed>> getFeeds();

    @GET("reader/api/0/stream/contents/user/-/state/com.google/reading-list")
    Single<List<Item>> getItems(@Query("xt") List<String> excludeTarget, @Query("n") int max, @Query("ot") Long lastModified);

    @GET("reader/api/0/stream/contents/user/-/state/com.google/starred")
    Single<List<Item>> getStarredItems(@Query("n") int max);

    @GET("reader/api/0/stream/items/ids")
    Single<List<String>> getItemsIds(@Query("xt") String excludeTarget, @Query("s") String includeTarget, @Query("n") int max);

    @GET("reader/api/0/tag/list?output=json")
    Single<List<Folder>> getFolders();

    @FormUrlEncoded
    @POST("reader/api/0/edit-tag")
    Completable setItemsState(@Field("T") String token, @Field("a") String addAction, @Field("r") String removeAction, @Field("i") List<String> itemIds);

    @FormUrlEncoded
    @POST("reader/api/0/subscription/edit")
    Completable createOrDeleteFeed(@Field("T") String token, @Field("s") String feedUrl, @Field("ac") String action);

    @FormUrlEncoded
    @POST("reader/api/0/subscription/edit")
    Completable updateFeed(@Field("T") String token, @Field("s") String feedUrl, @Field("t") String title, @Field("a") String folderId, @Field("ac") String action);

    @FormUrlEncoded
    @POST("reader/api/0/edit-tag")
    Completable createFolder(@Field("T") String token, @Field("a") String tagName);

    @FormUrlEncoded
    @POST("reader/api/0/rename-tag")
    Completable updateFolder(@Field("T") String token, @Field("s") String folderId, @Field("dest") String newFolderId);

    @FormUrlEncoded
    @POST("reader/api/0/disable-tag")
    Completable deleteFolder(@Field("T") String token, @Field("s") String folderId);
}

File: api/src/main/java/com/readrops/api/services/freshrss/FreshRSSService.kt
package com.readrops.api.services.freshrss

import com.readrops.api.services.freshrss.adapters.FreshRSSUserInfo
import com.readrops.db.entities.Feed
import com.readrops.db.entities.Folder
import com.readrops.db.entities.Item
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.http.Body
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Query

interface FreshRSSService {

    @POST("accounts/ClientLogin")
    suspend fun login(@Body body: RequestBody?): ResponseBody

    @GET("reader/api/0/token")
    suspend fun getWriteToken(): ResponseBody

    @GET("reader/api/0/user-info")
    suspend fun userInfo(): FreshRSSUserInfo

    @GET("reader/api/0/subscription/list?output=json")
    suspend fun getFeeds(): List<Feed>

    @GET("reader/api/0/tag/list?output=json")
    suspend fun getFolders(): List<Folder>

    @GET("reader/api/0/stream/contents/user/-/state/com.google/reading-list")
    suspend fun getItems(
        @Query("xt") excludeTarget: List<String>?,
        @Query("n") max: Int,
        @Query("ot") lastModified: Long?
    ): List<Item>

    @GET("reader/api/0/stream/contents/user/-/state/com.google/starred")
    suspend fun getStarredItems(@Query("n") max: Int): List<Item>

    @GET("reader/api/0/stream/items/ids")
    suspend fun getItemsIds(
        @Query("xt") excludeTarget: String?,
        @Query("s") includeTarget: String?,
        @Query("n") max: Int
    ): List<String>

    @FormUrlEncoded
    @POST("reader/api/0/edit-tag")
    suspend fun setItemsState(
        @Field("T") token: String,
        @Field("a") addAction: String?,
        @Field("r") removeAction: String?,
        @Field("i") itemIds: List<String>
    )

    @FormUrlEncoded
    @POST("reader/api/0/subscription/edit")
    suspend fun createOrDeleteFeed(
        @Field("T") token: String,
        @Field("s") feedUrl: String,
        @Field("ac") action: String,
        @Field("a") folderId: String?
    )

    @FormUrlEncoded
    @POST("reader/api/0/subscription/edit")
    suspend fun updateFeed(
        @Field("T") token: String,
        @Field("s") feedUrl: String,
        @Field("t") title: String,
        @Field("a") folderId: String,
        @Field("ac") action: String
    )

    @FormUrlEncoded
    @POST("reader/api/0/edit-tag")
    suspend fun createFolder(@Field("T") token: String, @Field("a") tagName: String)

    @FormUrlEncoded
    @POST("reader/api/0/rename-tag")
    suspend fun updateFolder(
        @Field("T") token: String,
        @Field("s") folderId: String,
        @Field("dest") newFolderId: String
    )

    @FormUrlEncoded
    @POST("reader/api/0/disable-tag")
    suspend fun deleteFolder(@Field("T") token: String, @Field("s") folderId: String)

    companion object {
        const val END_POINT = "api/greader.php/"
    }
}